### DISCUSSION TITLE
I2C Read without write beforehand

### DISCUSSION THREAD

**Discussion by Totalis** [MrYsLab/telemetrix#26] [Q&A] [answered]
Hi,  Firstly thank you for this amazing library and all the work you are putting in to this.

I have been using an MCP3428 ADC for a project.  It communicates using I2C but there is only a single configuration register on the device. This means that there is no need for register selection and that any write commands are directly into the configuration register.  

So when you come to read the ADC and the read request writes a byte into the configuration register, it scraps the previous config and overwrites it with the new one.

So what to do: Is there any way we could have a 'read_without_write' flag on the I2C_read function?  I am going to look at the code this morning and see if I can make a first attempt at a solution.

Thanks

Greg

**Comment by MrYsLab** [comment]
Hi Greg,
I have no experience with the MCP3428 device, so I need to ask the following. Does reinitializing for each read cause any problems? I don't know if you are using the one-shot or continuous mode, but have you tried both modes?

**Reply by Totalis** [reply]
So I tested this just this afternoon: sending another config byte does not invalidate the reading, instead it just makes the 'read_ready' bit go false instead of true.  Sending another byte does restart the ADC reading process so if you time your I2C read request poorly it can limit the sample rate quite badly.

I have tried using both modes, it's basically hard wired that as soon as the I2C write bit is received, the 'read_ready' bit goes false.

**Reply by MrYsLab** [reply]
Thanks for the info. I created a branch called [i2c_read ](https://github.com/MrYsLab/Telemetrix4Arduino/tree/i2c_read)for Telemetrix4Arduino. It contains a [temporary workaround to not write the register](https://github.com/MrYsLab/Telemetrix4Arduino/blob/fab9b0f9a679e5f0b7064c631b1aa86b766a5c1e/examples/Telemetrix4Arduino/Telemetrix4Arduino.ino#L932).
To test if this works for you, in the Arduino IDE, select the Telemetrix4Arduino example, and then overwrite the file with [this version](https://raw.githubusercontent.com/MrYsLab/Telemetrix4Arduino/i2c_read/examples/Telemetrix4Arduino/Telemetrix4Arduino.ino) and upload it to the Arduino.
Please let me know if that works. If it does, I will add a new command on the Python side and update the sketch with the new command as well.

**Reply by MrYsLab** [reply]
Also, make sure to set the register to None in the Python i2c_read call.

**Reply by Totalis** [reply]
Awesome, thank you very much. I will test tomorrow AM and get back to you.

**Reply by MrYsLab** [reply]
To support the feature, I am proposing that the i2c_read API change to:
```
    def i2c_read(self, address, register, number_of_bytes,
                 callback=None, i2c_port=0,
                 write_register=True):
        """
        Read the specified number of bytes from the
        specified register for the i2c device.


        :param address: i2c device address

        :param register: i2c register (or None if no register
                                       selection is needed)

        :param number_of_bytes: number of bytes to be read

        :param callback: Required callback function to report
                         i2c data as a result of read command

       :param i2c_port: 0 = default, 1 = secondary

       :param write_register: If True, the register is written
                                       before read
                              Else, the write is suppressed


        callback returns a data list:

        [I2C_READ_REPORT, address, register, count of data bytes,
         data bytes, time-stamp]
```

So if you wish to perform a read without the write, the syntax might look like this:

```
my_board.i2c_read(83, None, 6, the_callback, write_register=False)
```

Please let me know this seems acceptable. 

If it is, I will generate a new release for Telemetrix4Arduino, Telemetrix4Esp8266, telemetrix, and telemetrix-aio.

**Reply by Totalis** [reply]
That fix has worked perfectly. The device now returns the expected data, as well as handling the 'conversion not ready' bit properly.

I think that syntax is a great choice, I am aware this is a niche use case so making the original functionality the default is sensible, especially given that this change will still work with older python code that is not aware of the new functionality :)

Thank you for such a speedy response. How can I support your efforts further? (Buy you a beer or similar)

Greg

**Comment by MrYsLab** [comment, marked as answer]
@Totalis I have published a new version (5.0.0) of Telemetrix4Arduino and Telemetrix4Esp8266 and telemetrix and telemetrix-aio. 
When you get a chance, please update both the server and client software and let me know if the changes work for you. Thanks.

**Reply by Totalis** [reply]
Tested the updated version today and everything worked perfectly.  Thank you so much for this.

**Reply by MrYsLab** [reply]
@Totalis Glad that it is all working. Thanks for your suggestion!
